File: app/src/main/java/com/example/aida64remote/data/HtmlLabelParser.kt
package com.example.aida64remote.data

import android.text.Html
import java.util.regex.Pattern

object HtmlLabelParser {
    private val labelSpan = Pattern.compile(
        """(?is)<span\s+[^>]*id\s*=\s*["'](Label\d+)["'][^>]*>(.*?)</span>""",
    )
    private val siLabelNear = Pattern.compile(
        """(?is)id\s*=\s*["']SI(\d+)["'][\s\S]{0,1200}?float:\s*left[^>]*>(.*?)</div>""",
    )

    fun parseLabels(html: String): Map<String, String> {
        val labels = linkedMapOf<String, String>()

        val labelMatcher = labelSpan.matcher(html)
        while (labelMatcher.find()) {
            val id = labelMatcher.group(1) ?: continue
            val text = decodeHtml(labelMatcher.group(2).orEmpty())
            if (text.isNotBlank()) {
                labels[id] = text
            }
        }

        val siMatcher = siLabelNear.matcher(html)
        while (siMatcher.find()) {
            val index = siMatcher.group(1) ?: continue
            val text = decodeHtml(siMatcher.group(2).orEmpty())
            if (text.isNotBlank()) {
                labels["SIV$index"] = text
                labels["SI$index"] = text
            }
        }

        return labels
    }

    private fun decodeHtml(raw: String): String {
        val withoutTags = raw
            .replace(Regex("(?is)<[^>]+>"), " ")
            .replace("&nbsp;", " ")
            .replace(Regex("\\s+"), " ")
            .trim()
        return Html.fromHtml(withoutTags, Html.FROM_HTML_MODE_LEGACY).toString().trim()
    }
}

File: AIDA64Remote-iOS/AIDA64Remote/Data/HtmlLabelParser.swift
import Foundation

enum HtmlLabelParser {
    private static let labelSpan = try! NSRegularExpression(
        pattern: #"(?is)<span\s+[^>]*id\s*=\s*[\"'](Label\d+)[\"'][^>]*>(.*?)</span>"#,
        options: []
    )
    private static let siLabelNear = try! NSRegularExpression(
        pattern: #"(?is)id\s*=\s*[\"']SI(\d+)[\"'][\s\S]{0,1200}?float:\s*left[^>]*>(.*?)</div>"#,
        options: []
    )
    private static let tagStrip = try! NSRegularExpression(pattern: #"(?is)<[^>]+>"#, options: [])
    private static let whitespace = try! NSRegularExpression(pattern: #"\s+"#, options: [])

    static func parseLabels(_ html: String) -> [String: String] {
        var labels: [String: String] = [:]
        let nsRange = NSRange(html.startIndex..<html.endIndex, in: html)

        labelSpan.enumerateMatches(in: html, options: [], range: nsRange) { match, _, _ in
            guard let match,
                  let idRange = Range(match.range(at: 1), in: html),
                  let textRange = Range(match.range(at: 2), in: html)
            else { return }
            let id = String(html[idRange])
            let text = decodeHTML(String(html[textRange]))
            if !text.isEmpty {
                labels[id] = text
            }
        }

        siLabelNear.enumerateMatches(in: html, options: [], range: nsRange) { match, _, _ in
            guard let match,
                  let indexRange = Range(match.range(at: 1), in: html),
                  let textRange = Range(match.range(at: 2), in: html)
            else { return }
            let index = String(html[indexRange])
            let text = decodeHTML(String(html[textRange]))
            if !text.isEmpty {
                labels["SIV\(index)"] = text
                labels["SI\(index)"] = text
            }
        }

        return labels
    }

    private static func decodeHTML(_ raw: String) -> String {
        var text = tagStrip.stringByReplacingMatches(
            in: raw,
            options: [],
            range: NSRange(raw.startIndex..<raw.endIndex, in: raw),
            withTemplate: " "
        )
        text = text
            .replacingOccurrences(of: "&nbsp;", with: " ")
            .replacingOccurrences(of: "&amp;", with: "&")
            .replacingOccurrences(of: "&lt;", with: "<")
            .replacingOccurrences(of: "&gt;", with: ">")
            .replacingOccurrences(of: "&quot;", with: "\"")
            .replacingOccurrences(of: "&#39;", with: "'")
        text = whitespace.stringByReplacingMatches(
            in: text,
            options: [],
            range: NSRange(text.startIndex..<text.endIndex, in: text),
            withTemplate: " "
        )
        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
